docs/pages/machine-id/guides/gitlab.mdx

```diff
@@ -0,0 +1,166 @@
+---
+title: Using Machine ID with GitLab CI
+description: How to use Machine ID to SSH into Teleport nodes from GitLab CI
+---
+
+<Details
+  title="Version warning"
+  opened={true}
+  scope={["oss", "enterprise"]}
+  scopeOnly={true}
+  min="12.2"
+>
+  Machine ID for GitLab is available starting from Teleport `v12.2`.
+</Details>
+
+In this guide, you will use Teleport Machine ID to allow a GitLab pipeline to
+securely connect to a Teleport SSH node without the need for long-lived secrets.
+
+Machine ID for GitLab works with GitLab's cloud-hosted option and with
+self-hosted GitLab installations. **The minimum supported GitLab version is
+15.7**.
+
+This mitigates the risk of long-lived secrets such as passwords or SSH private
+keys being exfiltrated from your GitLab organization and provides many of
+the other benefits of Teleport such as auditing and finely-grained access
+control.
+
+## Prerequisites
+
+(!docs/pages/includes/edition-prereqs-tabs.mdx!)
+
+(!docs/pages/includes/tctl.mdx!)
+
+- A running instance of the Teleport SSH Service that you have registered with
+your Teleport cluster. For instructions on setting this up, see the
+[Getting Started Guide](../../server-access/introduction.mdx). The SSH node must
+include a user you want to grant access to. In this guide, we will call the SSH
+node `my-node` and the user `ci-user`. Replace these with values appropriate to
+your setup.
+- A GitLab project you wish to use. This can either be on GitLab's cloud-hosted
+offering (gitlab.com) or on a self-hosted GitLab instance. **When using a
+self-hosted GitLab instance, your Teleport Auth Server must be able to connect
+to your GitLab instance and your GitLab instance must be configured with a valid
+TLS certificate.**
+
+## Step 1/3. Create a join token
+
+In order to allow your GitLab CI to authenticate to your Teleport cluster,
+you'll first need to create a join token. These tokens define the rules that
+allow a bot to join your Teleport cluster.
+
+A GitLab join token contains allow rules that describe which pipelines should
+be able to use that token in order to join the Teleport cluster. A rule can
+contain multiple constraints, and any pipeline that matches all of the
+constraints within a single rule is granted access.
+
+The following constraints exist:
+- `sub`: a string uniquely identifying the CI run's source. It follows the
+  following format:
+  `project_path:{group}/{project}:ref_type:{type}:ref:{branch_name}`
+- `ref`: the git reference that triggered the CI run
+- `ref_type`: the type of reference that triggered the CI run. This will either
+  be `branch` or `tag`
+- `namespace_path`: the group or user that the project belongs to.
+- `project_path`: the full name of the project that the run belongs to. For
+example: `my-user/my-project`
+- `pipeline_source`: the trigger of the CI run. For example: `web`
+- `environment`: the environment that the job is associated with
```

Comment:
It looks like everything under `sub` is an explanation of the keys used within? If so, should we indent like this?

```suggestion
- `sub`: a string uniquely identifying the CI run's source. It follows the
  following format:
  `project_path:{group}/{project}:ref_type:{type}:ref:{branch_name}`
  - `ref`: the git reference that triggered the CI run
  - `ref_type`: the type of reference that triggered the CI run. This will either
    be `branch` or `tag`
  - `namespace_path`: the group or user that the project belongs to.
  - `project_path`: the full name of the project that the run belongs to. For
  example: `my-user/my-project`
  - `pipeline_source`: the trigger of the CI run. For example: `web`
  - `environment`: the environment that the job is associated with
```

If not, (and maybe still if so) where all these values, including `sub` go? I don't see it used in anything below.

Reply:
It's a weird one, all of these are constraints that can be used within an allow rule - but - the `sub` is a string made up of several of the other values. This is a design choice by Gitlab itself so it's a little difficult to explain their reasoning for it.

Reply:
The reason we don't see any of these used below is because we only use `project_path` in this tutorial. I included them all because these are valid options that a user configuring for their own environment way want to use to create tighter rules and because I want these values listed somewhere in the docs. If you feel this detracts from the simplicity of this tutorial, these could be pulled out into their own reference page that explains the entire GitLab joining method configuration options in detail.